main.go

```diff
@@ -255,6 +271,13 @@ func (svr *bridge) handleCall(w http.ResponseWriter, r *http.Request) {
 				log.Printf("    Alert %d", idx)
 			}
 
+			if alert.ExternalURL != "" {
```

Comment:
See prior note - I don't think ExternalURL will ever be sent, but rather GeneratorURL will be. That said, I do feel a little nervous that the prometheus template functions [use this value without checking that it is a non-nil pointer](https://github.com/prometheus/prometheus/blob/main/template/template.go#L342-L344).

Since I think it is possible to receive an alert that does not contain an ExternalURL field, it may not be safe to include this function.

Reply:
I didn't realize "ExternalURL" was exclusive to Grafana. After further testing, you are correct. It is exclusive to Grafana, but the result of no "ExternalURL" results in the same error when generating the template, so it doesn't cause any unexpected crashes. Still, that output could confuse an end user.
- error in Template: error executing template tmp: template: tmp:1:3: executing "tmp" at <externalURL>: error calling externalURL: runtime error: invalid memory address or nil pointer dereference

We could leave it as a Grafana alert exclusive? See code below.

```go
func renderTemplate(templateString string, data interface{}, externalURL *url.URL) (string, error) {
	var result string
	var err error
	var unsupportedFunc string

	//Excludes unsupported template function calls.
	switch {
	case (strings.Contains(templateString, "{{ query") || strings.Contains(templateString, "{{query")):
		unsupportedFunc = "query"
	case (strings.Contains(templateString, "{{ first") || strings.Contains(templateString, "{{first")):
		unsupportedFunc = "first"
	case (strings.Contains(templateString, "{{ label") || strings.Contains(templateString, "{{label")):
		unsupportedFunc = "label"
	case (strings.Contains(templateString, "{{ value") || strings.Contains(templateString, "{{value")):
		unsupportedFunc = "value"
	case (strings.Contains(templateString, "{{ strvalue") || strings.Contains(templateString, "{{strvalue")):
		unsupportedFunc = "strvalue"
	case (strings.Contains(templateString, "{{ safeHtml") || strings.Contains(templateString, "{{safeHtml")):
		unsupportedFunc = "safeHtml"
	case (strings.Contains(templateString, "{{ sortByLabel") || strings.Contains(templateString, "{{sortByLabel")):
		unsupportedFunc = "sortByLabel"
	case (strings.Contains(templateString, "{{ externalURL") || strings.Contains(templateString, "{{externalURL")) && fmt.Sprint(externalURL) == "<nil>":
		unsupportedFunc = "externalURL"
	case (strings.Contains(templateString, "{{ pathPrefix") || strings.Contains(templateString, "{{pathPrefix")) && fmt.Sprint(externalURL) == "<nil>":
		unsupportedFunc = "pathPrefix"
	default:
		unsupportedFunc = ""
	}

	if fmt.Sprint(externalURL) == "<nil>" {
		externalURL = nil
	}

	if unsupportedFunc == "" {
		titleTemplate := template.NewTemplateExpander(context.Background(), templateString, "tmp", data, 0, nil, externalURL, nil)
		result, err = titleTemplate.Expand()
		if err != nil {
			return "", fmt.Errorf("error in Template: %s", err)
		}
		return result, err
	} else if unsupportedFunc == "externalURL" || unsupportedFunc == "pathPrefix" {
		return "", fmt.Errorf("error in Template: The '%s' template function is only supported when the alert is generated through Grafana", unsupportedFunc)
	} else {
		return "", fmt.Errorf("error in Template: The bridge does not support the function %s", unsupportedFunc)
	}
}
```

Reply:
Since the module in actual use is the prometheus template library, maybe it's best to just say in the README that the functions available in Prometheus templating are available here as well - and provide a link to the [template.go](https://github.com/prometheus/prometheus/blob/main/template/template.go) file.

That way... 
* In the README, we could say something along the lines of, "Additional Prometheus template functions are available for the Grafana use case
* In this function we can just skip all logic for supported/unsupported functions (since a missing function would just be a regular template error)
* And we can ignore `ExternalURL` since it is Grafana-specific and we've called out that this is the Prometheus module

... said another way, we can just not care about what functions are available or not since the Template expansion should result in an error message indicating the same.

Thanks for testing the dereferenced `nil` failure scenario. This would normally be a runtime panic and crash, but since this is being executed in a Handler, I believe the mux wraps all calls with a `recover` to avoid the whole server going down if a bad request comes in and cannot handle.

Reply:
I looked a little more at the additional functions used in Prometheus. Some functions like `query` and `first` will not have the ability to pass a queryFunction (see template_test.go), so these function would not work.  Also, the few you already glanced at, like `label`, `value`, etc. rely on sample data, so those would not work either. 

This only leaves a few extra functions in this module that will integrate with the bridge correctly.

It looks like we have two options.

1. Leave the case statements listed, and say the bridge supports the functions listed in the Grafana Template Functions link. No additional explanation is required, and this could limit user confusion.

2. Let all functions pass with a link to the Grafana Template Functions and Template.go file. Some additional information would be required to explain all functions in the Grafana Template Function work. Add another section saying additional functions are available in the Template.go file, but some do not operate in the bridge due to data restrictions. These functions would need to be listed in our ReadMe.

Which option do you like best and would limit any user confusion?

Reply:
I think there's confusion about the point I'm making. This PR doesn't support Grafana functions. If you have a look at the `main.go` patch, you will see this import added:
> "github.com/prometheus/prometheus/template"

Then, later in the patch, you are using `template.NewTemplateExpander` - this function call is returning a new [template.Expander struct](https://github.com/prometheus/prometheus/blob/main/template/template.go#L146). This is all coming from https://github.com/prometheus/prometheus/blob/main/template/template.go - so at no point are any of the Grafana packages being used (thus, none of the grafana-only functions being available).

In fact, looking at the Grafana source code, it appears that [it is just using the same prometheus/template package](https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/state/template.go#L32-L64) as what your patch is using. This is why I was suggesting just referencing the function names that are in the Prometheus package rather than the Grafana docs - they (indirectly) are the same thing except a few functions added by Grafana.

The way the Grafana application adds custom functions is by [creating a FuncMap](https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/state/template_functions.go#L11-L15) and then [adds those functions](https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/state/template.go#L58) to the [standard FuncMap in the prometheus/template module](https://github.com/prometheus/prometheus/blob/main/template/template.go)... but since those are all Grafana-specific, it's still best to just point users to the documentation for the [Prometheus template functions](https://prometheus.io/docs/prometheus/latest/configuration/template_reference/) documentation.

Reply:
There is no confusion on this side regarding how things are called and functioning. I put this together by reading all the Prometheus calls and Grafana to see how they all worked. Initially, I set up a single funcMap with reReplaceAll and got it working, and I realized I should add all the supported functions, which put me down the path of adding in the template.go file.

The reason I put the two options is more for end-users trying to use the bridge and know what and what doesn't work. Since the Go Templating is focused around Grafana in the ReadMe, I was thinking it would be best to show that the Grafana Template functions work alongside the templating without adding any extra overhead confusion around the few additional functions in the template.go file.

I can see where you are coming from in wanting to keep things in-line for reference. I think this falls toward the developer's side the most to explain how things function and what additional features are being used. Let me see what I can put together for documentation in the ReadMe to try and keep the explanation high-level for the general user, but advance comments for anyone doing development or wants to try and use the few additional functions.